client/mainmenu/CHighScoreScreen.cpp

```diff
@@ -0,0 +1,401 @@
+				entry->Struct() = node.Struct();
+			}
+
+			Settings entry = persistentStorage.write["highscore"][calc.isCampaign ? "campaign" : "scenario"][std::to_string(i)]["player"];
+			entry->String() = text;
+			if(calc.isCampaign)
+			{
+				Settings entry2 = persistentStorage.write["highscore"]["campaign"][std::to_string(i)]["campaign"];
+				entry2->String() = calc.calculate().cheater ? CGI->generaltexth->translate("core.genrltxt.260") : calc.parameters[0].campaign;
+			}
+			else
+			{
+				Settings entry3 = persistentStorage.write["highscore"]["scenario"][std::to_string(i)]["land"];
+				entry3->String() = calc.calculate().cheater ? CGI->generaltexth->translate("core.genrltxt.260") : calc.parameters[0].land;
+			}
+			Settings entry4 = persistentStorage.write["highscore"][calc.isCampaign ? "campaign" : "scenario"][std::to_string(i)]["days"];
+			entry4->Integer() = calc.calculate().sumDays;
+			Settings entry5 = persistentStorage.write["highscore"][calc.isCampaign ? "campaign" : "scenario"][std::to_string(i)]["points"];
+			entry5->Integer() = calc.calculate().cheater ? 0 : calc.calculate().total;
+
+			Settings entry6 = persistentStorage.write["highscore"][calc.isCampaign ? "campaign" : "scenario"][std::to_string(i)]["datetime"];
+			entry6->String() = vstd::getFormattedDateTime(std::time(0));
+
+			return i;
+		}
+	}
+
+	return -1;
+}
+
+void CHighScoreInputScreen::show(Canvas & to)
+{
+	CCS->videoh->update(pos.x, pos.y, to.getInternalSurface(), true, false,
+	[&]()
+	{
+		if(won)
+		{
+			CCS->videoh->close();
+			video = "HSLOOP.SMK";
+			CCS->videoh->open(VideoPath::builtin(video));
+		}
+		else
+			close();
+	});
+	redraw();
+
+	CIntObject::show(to);
+}
+
+void CHighScoreInputScreen::activate()
+{
+	if(!CCS->videoh->open(VideoPath::builtin(video)))
+	{
+		if(!won)
+			close();
+	}
+	else
+		background = nullptr;
+	CIntObject::activate();
+}
+
+void CHighScoreInputScreen::deactivate()
+{
+	CCS->videoh->close();
+	CIntObject::deactivate();
+}
+
+void CHighScoreInputScreen::clickPressed(const Point & cursorPosition)
+{
+	OBJ_CONSTRUCTION_CAPTURING_ALL_NO_DISPOSE;
+
+	if(!won)
+	{
+		close();
+		return;
+	}
+
+	if(!input)
+	{
+		input = std::make_shared<CHighScoreInput>(
+		[&] (std::string text) 
+		{
+			if(!text.empty())
+			{
+				int pos = addEntry(text);
+				close();
+				GH.windows().createAndPushWindow<CHighScoreScreen>(calc.isCampaign ? CHighScoreScreen::HighScorePage::CAMPAIGN : CHighScoreScreen::HighScorePage::SCENARIO, pos);
+			}
+			else
+				close();
+		});
+	}
+}
+
+void CHighScoreInputScreen::keyPressed(EShortcut key)
+{
+	clickPressed(Point());
+}
+
+CHighScoreInput::CHighScoreInput(std::function<void(std::string text)> readyCB)
+	: CWindowObject(0), ready(readyCB)
+{
+	OBJ_CONSTRUCTION_CAPTURING_ALL_NO_DISPOSE;
+
+	pos = center(Rect(0, 0, 232, 212));
+	updateShadow();
+
+	background = std::make_shared<CPicture>(ImagePath::builtin("HIGHNAME"));
```

Comment:
For background you can provide it into CWindowObject constuctor:
`CWindowObject(0, ImagePath::builtin("HIGHNAME")),`
It will automaticallly load this image and update `pos` of your widget using background dimensions